src/pyggdrasil/tree_inference/_tree.py

```diff
@@ -0,0 +1,273 @@
+    Used boolean matrix exponentiation to find descendants.
+
+    Complexity:
+        Naive: O(n^3 * (n-1)) where n is the number of nodes in the tree including root.
+        TODO: - Consider implementing 'Exponentiation by Squaring Algorithm'
+                    for  O(n^3 * log(m)
+              - fix conditional exponentiation for exponent < n-1
+    Args:
+    - adj_matrix: a JAX array of shape (n, n) representing the adjacency matrix
+    - labels: a JAX array of shape (n,) representing the labels of the nodes
+    - parent: an integer representing
+        the label of the node whose descendants we want to find
+
+    Returns:
+    - a JAX array of integers representing the labels of the descendants of node parent
+      in order of nodes in the adjacency matrix, i.e. the order of the labels
+      if includeParent is True, the parent is included in the list of descendants
+    """
+    # get number of nodes
+    n = adj_matrix.shape[0]
+    # get ancestor matrix
+    ancestor_mat = _get_ancestor_matrix(adj_matrix, n)
+    # get index of parent
+    parent_idx = int(jnp.where(labels == parent)[0])
+    # get descendants
+    desc = jnp.where(ancestor_mat[parent_idx, :])[0]
+    # get labels correspond to indices
+    desc_labels = labels[desc]
+    # remove parent - as self-looped
+    if not includeParent:
+        desc_labels = desc_labels[desc_labels != parent]
+    return desc_labels
+
+
+def _expon_adj_mat(adj_matrix: Array, exp: int, cond: bool = False):
+    """Exponentiation of adjacency matrix.
+
+    Complexity: O(n^3 * m) where n is the size of the square matrix and m the exponent
+            if cond= True the 'm' < n-1 hence speedup
+
+    TODO: Consider implementing 'Exponentiation by Squaring Algorithm'
+            for  O(n^3 * log(m)
+
+    Args:
+        adj_matrix : adjacency matrix
+        exp: exponent
+        cond: if to use conditional exponentiation,
+                stop if matrix does not change anymore
+    """
+    bool_mat = jnp.where(adj_matrix == 1, True, False)
+    adj_matrix_exp = bool_mat
+    if not cond:
+
+        def body(carry, _):
+            return jnp.dot(carry, bool_mat), None
+
+        adj_matrix_exp = jax.lax.scan(body, bool_mat, jnp.arange(exp))[0]
+    elif cond:
+        # TODO: fix this - this is not working
+        raise NotImplementedError("Conditional exponentiation not implemented yet.")
+
+        exp_counter = 0
+
+        def loop_cond_fn(carry):
+            prev_matrix, curr_matrix = carry
+            nonlocal exp_counter
+            exp_counter = exp_counter + 1
+            return ~(jnp.array_equal(prev_matrix, curr_matrix)) and (exp_counter <= exp)
+
+        def loop_body_fn(carry):
+            prev_matrix, curr_matrix = carry
+            new_matrix = jnp.dot(curr_matrix, bool_mat)
+            return curr_matrix, new_matrix
+
+        (_, adj_matrix_exp), _ = jax.lax.while_loop(
+            loop_cond_fn, loop_body_fn, (bool_mat, bool_mat)
+        )
```

Comment:
A failed attempt to stop exponentiation if no changes occurred, in the last step. - may remove.

Reply:
Removed.